docs/code-guidelines.md

```diff
@@ -0,0 +1,1141 @@
+# Contributing
+
+If you want to contribute to a project and improve it, your help is welcome. We want to make Gaia as good as it can be. Contributing is also a great way to learn more about blockchain technology and improve it. Please read this document and follow our guidelines to make the process as smooth as possible. We are happy to review your code but please ensure that you have a reasonable and clean pull request.
+
+This documents idiomatic conventions in the Go code that we follow at Uber. A lot of these are general guidelines for Go, while others extend upon external resources:
+
+1. [Effective Go](https://golang.org/doc/effective_go.html)
+2. [Go Common Mistakes](https://github.com/golang/go/wiki/CommonMistakes)
+3. [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
+
+## How to make a clean pull request
+
+Look for a project's contribution instructions and follow them:
+
+- Create a personal fork of the project on GitHub.
+- Clone the fork on your local machine. Your remote repo on Github is called `origin`.
+- Add the original repository as a remote called `upstream`.
+- If you created your fork a while ago, pull upstream changes should be into your local repository.
+- Create a new branch to work on! Start your branch from `main`.
+- Implement/fix your feature, and comment on your code.
+- Follow the code style of the project, including indentation. Use `gofmt` and `goimports` if you are unsure.
+- Run the tests
+- Write or adapt tests as needed.
+- Add or change the documentation and comments as needed.
+- Write your commit messages in the present tense. Your commit message should describe what the commit, when applied, does to the code, not what you did to the code.
+- Push your branch to your fork on GitHub, the remote `origin`.
+- Avoid push force after the review already start. The commit history should tell a narrative.
+- From your fork, open a pull request in the correct branch. Target the project's `main` branch.
+- If we request further changes, push them to your branch. The PR will be updated automatically.
+- Once the pull request is approved and merged, you can pull the changes from `upstream` to your local repo and delete your extra branch(es).
+
+Is it not uncommon for a PR to accumulate commits and merges with time. Gaia is in constant change. If your PR falls out of sync with the upstream `main`, you need to rebase. We can't reliably review code that is spread over too many other changes to the codebase.
+
+## Run tests
+
+- Run unit tests
+
+```shell
+make test-unit
+```
+
+- Run the unit tests and output the coverage file (coverage.txt).
+
+```shell
+make test-unit-cover
+```
+
+- Run the unit tests with the race condition flag on.
+
+```shell
+make test-race
+```
+
+- Run end-to-end integration tests (Docker needed).
+
+```shell
+make docker-build-hermes && \
```

Comment:
isn't hermes docker built as part of `make docker-build-debug` ?

Reply:
it is not part of `make-docker-build-debug`. and we are not using this hermes image for e2e test actually? e2e is using a docker image [here](https://github.com/cosmos/gaia/blob/0b2781064be85a96c113776ba119df5084bef8c0/tests/e2e/e2e_ibc_test.go#L45).